api/src/main/java/marquez/db/JobDao.java

```diff
@@ -91,6 +78,53 @@
   @SqlQuery("SELECT COUNT(*) FROM jobs")
   int count();
 
+  default JobRow upsert(
+      NamespaceName namespaceName, JobName jobName, JobMeta jobMeta, ObjectMapper mapper) {
+    Instant createdAt = Instant.now();
+    NamespaceRow namespace =
+        createNamespaceDao()
+            .upsert(
+                UUID.randomUUID(), createdAt, namespaceName.getValue(), DEFAULT_NAMESPACE_OWNER);
+    JobContextRow contextRow =
+        createJobContextDao()
+            .upsert(
+                UUID.randomUUID(),
+                createdAt,
+                Utils.toJson(jobMeta.getContext()),
+                Utils.checksumFor(jobMeta.getContext()));
+
+    return upsert(
+        UUID.randomUUID(),
+        jobMeta.getType(),
+        createdAt,
+        namespace.getUuid(),
+        namespace.getName(),
+        jobName.getValue(),
+        jobMeta.getDescription().orElse(null),
+        contextRow.getUuid(),
+        toUrlString(jobMeta.getLocation().orElse(null)),
+        toJson(jobMeta.getInputs(), mapper),
+        toJson(jobMeta.getOutputs(), mapper));
+  }
+
+  default String toUrlString(URL url) {
+    if (url == null) {
+      return null;
+    }
+    return url.toString();
+  }
+
+  default PGobject toJson(Set<DatasetId> dataset, ObjectMapper mapper) {
+    try {
+      PGobject jsonObject = new PGobject();
+      jsonObject.setType("json");
+      jsonObject.setValue(mapper.writeValueAsString(dataset));
+      return jsonObject;
+    } catch (Exception e) {
+      return null;
```

Comment:
This just tracks what job inputs/outputs the client expected to have for their job but we don't actually use this data for anything. I'd rather just suppress any exceptions or not include it all rather than fail.

Reply:
It might be helpful to log a warning?

Reply:
Dao classes can't attach an slf4j logger